### PR TITLE
Add read/write functions for null values + delete HDU

### PR DESCRIPTION
**Pull request by jishnub**
This PR adds several functions:

- fits_delete_hdu
- fits_read_pixnull
- fits_write_pixnull
- fits_write_subset
- fits_write_null_img

Plus it adds explicit checks for closed files so that segfaults are avoided. This changes the present behavior, but IMO errors are better than segfaults. Various other tests and docstring improvements are added as well.

### CODE REVIEW

File: .github/workflows/TagBot.yml [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 jobs:
   TagBot:
+    if: github.event_name == 'workflow_dispatch' || github.actor == 'JuliaTagBot'
     runs-on: ubuntu-latest
     steps:
       - uses: JuliaRegistries/TagBot@v1
         with:
           token: ${{ secrets.GITHUB_TOKEN }}
-          ssh: ${{ secrets.DOCUMENTER_KEY }}
+          ssh: ${{ secrets.DOCUMENTER_KEY }}
```
Comment (giordano) [inline .github/workflows/TagBot.yml:15]:
```suggestion
          ssh: ${{ secrets.DOCUMENTER_KEY }}

```

File: src/CFITSIO.jl [hunk 2]
```diff
@@ -61,13 +63,19 @@ export FITSFile,
...
+    fits_write_null_img,
     fits_write_record,
     fits_write_tdim,
     libcfitsio_version,
     cfitsio_typecode,
     bitpix_from_type,
     type_from_bitpix
 
+const READONLY = 0
+const READWRITE = 1
```
Comment (giordano) [inline src/CFITSIO.jl:77]: Maybe these can be `Enum`s?  That'd allow us to make the signature more specific, too

File: src/CFITSIO.jl [hunk 3]
```diff
@@ -570,8 +629,9 @@ function fits_write_key(
     f::FITSFile,
     keyname::String,
     value::Union{Real,String},
-    comment::String,
-)
+    comment::String)
```
Comment (giordano) [inline src/CFITSIO.jl:632]: The goal of this was probably to allow adding more arguments without disrupting git blame, which is very valuable.  I'd leave it as it was before
```suggestion
    comment::String,
)
```

File: src/CFITSIO.jl [hunk 4]
```diff
@@ -1043,16 +1375,16 @@ function fits_read_subset(
         (
             Ptr{Cvoid},
             Cint,
-            Ptr{Clong},
-            Ptr{Clong},
-            Ptr{Clong},
+            Ref{Clong},
+            Ref{Clong},
+            Ref{Clong},
```
Comment (giordano) [inline src/CFITSIO.jl:1380]: Why this change?

**Review by giordano — APPROVED** [review]
It looks overall good, thanks!  Did you observe a measurable performance impact due to the extra checks?  I wouldn't expect large regressions.

@mileslucas any chance of having a look at this?

> but IMO errors are better than segfaults

Agreed!

File: src/CFITSIO.jl (hunk 2 above)
Comment (mileslucas) [reply src/CFITSIO.jl:77]: I agree an enum seems more appropriate

**Comment by mileslucas** [comment]
Overall I agree with Mosè, as long as there's no big slowdowns or gotchas, this seems like a massive improvement, thanks!

File: src/CFITSIO.jl (hunk 4 above)
Comment (jishnub) [reply src/CFITSIO.jl:1380]: My understanding was that `Ref` is equivalent to `Ptr` except it's an object that has its memory managed by Julia, so there are no risks of memory being freed during the `ccall`. Is this not a correct usage of `Ref`? Or is it unnecessary as these are arguments to the function anyway?

File: src/CFITSIO.jl (hunk 3 above)
Comment (jishnub) [reply src/CFITSIO.jl]: I've fixed this in 031d41b

File: src/CFITSIO.jl (hunk 2 above)
Comment (jishnub) [reply src/CFITSIO.jl]: I've changed this to use an enum in 031d41b. What do you recommend the signature should be? Currently it accepts ints as well as `READONLY/READWRITE` types.

Comment (mileslucas) [reply src/CFITSIO.jl]: I think it might look better to do something like
```julia
@enum FileMode R=0 RW=1
function fits_open_file(filename, mode=R)
```
If we want to restrict users to only using the enum, not any integer, we just have to type `mode` to be `mode::FileMode=R`. I like this idea but I realize it might be a nuisance, thoughts?

I like the briefer names because they won't be exported and its nicer to call `CFITSIO.R` and `CFITSIO.RW`.

File: src/CFITSIO.jl (hunk 4 above)
Comment (mileslucas) [reply src/CFITSIO.jl:1386]: From the Julia manual:
> When passed as a `ccall` argument (either as a `Ptr` or `Ref` type), a Ref object will be converted to a native pointer to the data it references. For most T, or when converted to a `Ptr{Cvoid}`, this is a pointer to the object data. When T is an `isbits` type, this value may be safely mutated, otherwise mutation is strictly undefined behavior.

So at first glance it doesn't seem like there is a difference between `Ptr` and `Ref`? I'm not sure, though.

**Comment by giordano** [comment]
I just noted there is this pull request still pending.  Is there anything still do besides some benchmarks to make sure we aren't making everything horribly slower?

**Comment by jishnub** [comment]
Yes, I imagine small fixes are needed here and I don't think it'll slow things down much. I will get to this in a couple of days

File: src/CFITSIO.jl (hunk 2 above)
Comment (jishnub) [reply src/CFITSIO.jl]: I think restricting users to `FileMode` might be breaking, as integers should be valid arguments as well. Perhaps `Union{FileMode, Integer}` might be more appropriate. However I wonder if the specialization is worth it, since we explicitly convert the argument to `Int` anyway.

Regarding the shorter names, I wonder if readability is increased by the more verbose ones? Although perhaps `R` and `RW` would be familiar from their use in `FITSIO`, so that seems reasonable too.

**Comment by jishnub** [comment]
It appears that certain allocations may be avoided in the usual cases. After this certain functions may actually become marginally faster.

On master:
```julia
julia> using CFITSIO
[ Info: Precompiling CFITSIO [3b1b4be9-1499-4b22-8d78-7db3344d1961]

julia> using BenchmarkTools

julia> dir = tempdir()
"/tmp"

julia> filename = joinpath(dir, "temp.fits")
"/tmp/temp.fits"

julia> f = CFITSIO.fits_clobber_file(filename);

julia> a = ones(2,2);

julia> fits_create_img(f, eltype(a), [size(a)...]);

julia> fits_write_pix(f, a);

julia> close(f)

julia> b = similar(a);

julia> f = fits_open_file(filename, 0);

julia> @btime fits_read_pix($f, $b);
  318.987 ns (2 allocations: 192 bytes)
```

This PR:
```julia
julia> using CFITSIO
[ Info: Precompiling CFITSIO [3b1b4be9-1499-4b22-8d78-7db3344d1961]

julia> using BenchmarkTools

julia> dir = tempdir()
"/tmp"

julia> filename = joinpath(dir, "temp.fits")
"/tmp/temp.fits"

julia> f = CFITSIO.fits_clobber_file(filename);

julia> a = ones(2,2);

julia> fits_create_img(f, eltype(a), [size(a)...]);

julia> fits_write_pix(f, a);

julia> close(f)

julia> b = similar(a);

julia> f = fits_open_file(filename, 0);

julia> @btime fits_read_pix($f, $b);
  307.530 ns (1 allocation: 96 bytes)
```

**Comment by jishnub** [comment]
bump

**Comment by kbarbary** [comment]
> Plus it adds explicit checks for closed files so that segfaults are avoided.

Just a historical note here: Originally when CFITSIO was in the same repo as FITSIO, the idea was to have CFITSIO be mostly one-to-one wrappers around the c functions. Higher-level functionality, like composing multiple function calls in order to avoid segfaults went into FITSIO. This was OK because it was understood that CFITSIO was a "low-level" wrapper and mostly intended for use only by FITSIO itself.

There are a lot of "is file open?" checks performed in FITSIO. If these checks are moved "down the stack" to CFITSIO, you may want to check that there aren't redundant checks being performed in FITSIO.

**Comment by kbarbary** [comment]
> Plus it adds explicit checks for closed files so that segfaults are avoided.

Just a historical note here: Originally when CFITSIO was in the same repo as FITSIO, the idea was to have CFITSIO be mostly one-to-one wrappers around the c functions. Higher-level functionality, like composing multiple function calls in order to avoid segfaults went into FITSIO. This was OK because it was understood that CFITSIO was a "low-level" wrapper and mostly intended for use only by FITSIO itself.

There are a lot of "is file open?" checks performed in FITSIO. If these checks are moved "down the stack" to CFITSIO, you may want to check that there aren't redundant checks being performed in FITSIO.

**Comment by jishnub** [comment]
That's a good point (and I think currently there are such tests), but I wonder if the cost associated with such redundant tests are significant compared to other IO tasks?
